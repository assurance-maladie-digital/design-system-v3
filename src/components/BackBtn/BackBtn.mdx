import {Controls, Canvas, Meta, Source, Story} from '@storybook/blocks';
import * as BackBtnStories from './BackBtn.stories';

<Meta of={BackBtnStories} />

# BackBtn

L’élément `BackBtn` est utilisé pour permettre à l’utilisateur de revenir à la page précédente.

<Story of={BackBtnStories.Info} />

<Canvas of={BackBtnStories.Default}  />

# API

<Controls of={BackBtnStories.Default} />

<Source dark code={`
<script setup lang="ts">
import BackBtn from '@/components/BackBtn/BackBtn.vue'
</script>

<template>
    <BackBtn/>
</template>
`} />