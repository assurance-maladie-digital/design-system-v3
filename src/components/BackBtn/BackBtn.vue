<script lang="ts" setup>
	import { ref, computed } from 'vue'
	import { VIcon, VBtn } from 'vuetify/components'
	import { mdiArrowLeft } from '@mdi/js'
	import { locales } from './locales'

	const props = defineProps<{
		hideBackIcon?: boolean
		dark?: boolean
	}>()

	const backIcon = ref(mdiArrowLeft)
	const isDark = computed(() => props.dark ?? false)
	const iconColor = computed(() => isDark.value ? 'white' : 'primary')
	const buttonVariant = computed(() => isDark.value ? 'outlined' : 'text')
	const buttonTheme = computed(() => isDark.value ? 'dark' : 'light')
	const buttonColor = computed(() => isDark.value ? 'white' : 'primary')

	const buttonClasses = computed(() => ({
		'px-0': !isDark.value,
		'pr-1': !isDark.value && !props.hideBackIcon,
	}))

</script>

<template>
	<VBtn
		v-bind="$attrs"
		:variant="buttonVariant"
		:theme="buttonTheme"
		:color="buttonColor"
		:outlined="isDark"
		:class="['vd-back-btn', 'text-none', buttonClasses]"
	>
		<slot name="icon">
			<VIcon
				v-if="!props.hideBackIcon"
				:color="iconColor"
				:class="{ 'ml-n1': isDark }"
				class="mr-1"
			>
				{{ backIcon }}
			</VIcon>
		</slot>

		<slot>
			{{ locales.label }}
		</slot>
	</VBtn>
</template>

<style lang="scss" scoped>
// Désactiver l'état de hover sur le thème clair
.v-btn.v-theme--light:deep() {
  .v-btn__underlay,
  .v-btn__overlay {
    display: none;
  }
}
</style>
