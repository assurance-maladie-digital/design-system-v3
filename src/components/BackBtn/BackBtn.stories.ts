import type { Meta, StoryObj } from '@storybook/vue3'

import BackBtn from './BackBtn.vue'
import SyAlert from '@/components/SyAlert/SyAlert.vue'

import { VSheet } from 'vuetify/components'

const meta = {
	title: 'Composants/Boutons/BackBtn',
	component: BackBtn,
	parameters: {
		layout: 'fullscreen',
	},
	argTypes: {
		dark: {
			control: 'boolean',
		},
		hideBackIcon: {
			control: 'boolean',
		},
	},
} as Meta<typeof BackBtn>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<VSheet class="pa-4">
		<BackBtn />
	</VSheet>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import BackBtn from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		dark: false,
	},
	render: (args) => {
		return {
			components: { BackBtn, VSheet },
			setup() {
				return { args }
			},
			template: `
				<VSheet
					v-if="args.dark"
					color="primary"
					class="pa-4"
				>
				<BackBtn v-bind="args" />
				</VSheet>
				<VSheet
					v-else
					class="pa-4"
				>
					<BackBtn v-bind="args" />
				</VSheet>
			`,
		}
	},
}

export const Dark: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<VSheet color="primary" class="pa-4">
		<BackBtn dark />
	</VSheet>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import BackBtn from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		dark: true,
	},
	render: (args) => {
		return {
			components: { BackBtn, VSheet },
			setup() {
				return { args }
			},
			template: `
				<VSheet
					color="primary"
					class="pa-4"
				>
				<BackBtn v-bind="args" />
				</VSheet>
			`,
		}
	},
}

export const HideBackIcon: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<VSheet class="pa-4">
		<BackBtn hide-back-icon />
	</VSheet>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import BackBtn from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		hideBackIcon: true,
	},
	render: (args) => {
		return {
			components: { BackBtn },
			setup() {
				return { args }
			},
			template: `
				<VSheet class="pa-4">
					<BackBtn v-bind="args" />
				</VSheet>
			`,
		}
	},
}

export const DarkAndHideBackIcon: Story = {
	parameters: {
		sourceCode: [
			{
				name: 'Template',
				code: `<template>
	<VSheet class="pa-4">
		<BackBtn
			dark 
			hide-back-icon 
		/>
	</VSheet>
</template>
				`,
			},
			{
				name: 'Script',
				code: `<script setup lang="ts">
	import BackBtn from '@cnamts/synapse'
</script>
				`,
			},
		],
	},
	args: {
		dark: true,
		hideBackIcon: true,
	},
	render: (args) => {
		return {
			components: { BackBtn },
			setup() {
				return { args }
			},
			template: `
				<VSheet
					color="primary"
					class="pa-4"
				>
				<BackBtn v-bind="args" />
				</VSheet>
			`,
		}
	},
}

export const Info: Story = {
	render: (args) => {
		return {
			components: { SyAlert },
			setup() {
				return { args }
			},
			template: `
				<SyAlert v-model="args.modelValue" :type="args.type" :variant="tonal" :closable="false">
					<template #default>Par défaut, le composant BackBtn n'effectue aucune action, vous devez
						implémenter une redirection ou une action lors du clic sur le bouton.
					</template>
				</SyAlert>
			`,
		}
	},
	tags: ['!dev'],
}
