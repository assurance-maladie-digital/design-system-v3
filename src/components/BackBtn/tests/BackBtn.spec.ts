import { mount } from '@vue/test-utils'
import { expect, describe, it } from 'vitest'
import BackBtn from '@/components/BackBtn/BackBtn.vue'
import { mdiArrowLeft } from '@mdi/js'
import { vuetify } from '@tests/unit/setup'

describe('BackBtn.vue', () => {
	it('render correctly', async () => {
		const wrapper = mount(BackBtn, {
			global: {
				plugins: [vuetify],
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})
	it('renders with default props', () => {
		const wrapper = mount(BackBtn)
		expect(wrapper.find('.vd-back-btn').exists()).toBe(true)
		expect(wrapper.find('VIcon').text()).toBe(mdiArrowLeft)
		expect(wrapper.find('VIcon').attributes('color')).toBe('primary')
	})

	it('renders with dark theme', () => {
		const wrapper = mount(BackBtn, {
			props: { dark: true },
		})
		expect(wrapper.find('VIcon').attributes('color')).toBe('white')
	})

	it('hides back icon when hideBackIcon is true', () => {
		const wrapper = mount(BackBtn, {
			props: { hideBackIcon: true },
		})
		expect(wrapper.find('VIcon').exists()).toBe(false)
	})

	it('applies correct button classes based on props', () => {
		const wrapper = mount(BackBtn, {
			props: { dark: false, hideBackIcon: false },
		})
		expect(wrapper.find('.vd-back-btn').classes()).toContain('pr-1')
	})

	it('applies correct button classes when dark is true and hideBackIcon is false', () => {
		const wrapper = mount(BackBtn, {
			props: { dark: true, hideBackIcon: false },
		})
		expect(wrapper.find('.vd-back-btn').classes()).not.toContain('pr-1')
	})
	it('returns false when dark prop is not provided', () => {
		const wrapper = mount(BackBtn)
		expect(wrapper.vm.dark).toBe(false)
	})

	it('returns true when dark prop is true', () => {
		const wrapper = mount(BackBtn, {
			props: { dark: true },
		})
		expect(wrapper.vm.dark).toBe(true)
	})

	it('returns false when dark prop is false', () => {
		const wrapper = mount(BackBtn, {
			props: { dark: false },
		})
		expect(wrapper.vm.dark).toBe(false)
	})
})
