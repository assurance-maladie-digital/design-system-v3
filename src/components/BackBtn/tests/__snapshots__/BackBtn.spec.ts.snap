// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackBtn.vue > render correctly 1`] = `
"<button data-v-efb74141="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-text vd-back-btn text-none px-0 pr-1" outlined="false"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
  <!----><span class="v-btn__content" data-no-activator=""><i data-v-efb74141="" class="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z mdi v-icon notranslate v-theme--light v-icon--size-default text-primary mr-1" aria-hidden="true"></i>Retour</span>
  <!---->
  <!---->
</button>"
`;
